lib/src/rules/prefer_interpolation_to_compose_strings.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.prefer_interpolation_to_compose_strings;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Use interpolation to compose strings and values.';
+
+const _details = r'''
+
+**PREFER** using interpolation to compose strings and values.
+
+**BAD:**
+```
+'Hello, ' + name + '! You are ' + (year - birth) + ' years old.';
+```
+
+**GOOD:**
+```
+'Hello, $name! You are ${year - birth} years old.';
+```
+
+''';
+
+class PreferInterpolationToComposeStrings extends LintRule {
+  _Visitor _visitor;
+  PreferInterpolationToComposeStrings()
+      : super(
+            name: 'prefer_interpolation_to_compose_strings',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    if (node.operator.type == TokenType.PLUS_EQ &&
+        (node.leftHandSide.bestType.name == 'String' ||
+            node.rightHandSide.bestType.name == 'String')) {
```

Comment:
After you've added it, consider using `DartTypeUtilities.isClass(node.bestType, 'String', 'dart.core')` to reduce the chances of false positives.